File: lintcode/java/MergeSortedArray.java
class Solution {
    /**
     * @param A: sorted integer array A which has m elements, 
     *           but size of A is m+n
     * @param B: sorted integer array B which has n elements
     * @return: void
     */
    public void mergeSortedArray(int[] A, int m, int[] B, int n) {
        // parameter check
        if(m+n > A.length) throw new IllegalArgumentException("xxx");
        
        int i = m-1;
        int j = n-1;
        int k = A.length-1;
        
        while (i >= 0 && j >= 0){
            if(A[i] <= B[j]){
                A[k--] = B[j--];
            }else{
                A[k--] = A[i--];
            }
        }
        
        while(j >= 0){
            A[k--] = B[j--];
        }
    }
}

File: lintcode/cpp/Merge_Sorted_Array.cpp
class Solution {
public:
    /**
     * @param A and B: sorted integer array A and B.
     * @return: A new sorted integer array
     */
    vector<int> mergeSortedArray(vector<int> &A, vector<int> &B) {
        // write your code here
        int m = A.size();
        int n = B.size();
        int i = 0;
        int j = 0;
        
        vector<int> result;
        while(i < m && j < n){
            if(A[i] <= B[j]) result.push_back(A[i++]);
            else result.push_back(B[j++]);
        }
        
        while(i < m) result.push_back(A[i++]);
        while(j < n) result.push_back(B[j++]);
        
        return result;
    }
};
